// retire/lib/reporters/cyclonedx-json.js
"use strict";
/*jshint esversion: 6 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const retire = __importStar(require("../retire"));
const fs = __importStar(require("fs"));
const uuid_1 = require("uuid");
const utils_1 = require("./utils");
function configureCycloneDXJSONLogger(logger, writer, config, hash) {
    let vulnsFound = false;
    const finalResults = {
        version: retire.version,
        start: new Date().toISOString(),
        data: [],
        messages: [],
        errors: [],
    };
    logger.info = finalResults.messages.push;
    logger.debug = config.verbose
        ? finalResults.messages.push
        : function () {
            return;
        };
    logger.warn = logger.error = finalResults.errors.push;
    logger.logVulnerableDependency = function (finding) {
        vulnsFound = true;
        finalResults.data.push(finding);
    };
    logger.logDependency = function (finding) {
        if (finding.results.length > 0) {
            finalResults.data.push(finding);
        }
    };
    logger.close = function (callback) {
        const write = vulnsFound ? writer.err : writer.out;
        const seen = new Set();
        const components = finalResults.data
            .filter((d) => d.results)
            .map((r) => r.results
            .map((dep) => {
            dep.version = (dep.version.split('.').length >= 3 ? dep.version : dep.version + '.0').replace(/-/g, '.');
            let hashes;
            const filepath = r.file;
            if (filepath) {
                const file = fs.readFileSync(filepath);
                hashes = [
                    { alg: 'MD5', content: hash.md5(file) },
                    { alg: 'SHA-1', content: hash.sha1(file) },
                    { alg: 'SHA-256', content: hash.sha256(file) },
                    { alg: 'SHA-512', content: hash.sha512(file) },
                ];
            }
            const purl = (0, utils_1.generatePURL)(dep);
            if (seen.has(purl))
                return undefined;
            seen.add(purl);
            return {
                type: 'library',
                name: dep.component,
                version: dep.version,
                purl: purl,
                hashes: hashes,
            };
        })
            .filter((x) => x != undefined))
            .reduce((a, b) => a.concat(b), []);
        write(JSON.stringify({
            bomFormat: 'CycloneDX',
            specVersion: '1.4',
            serialNumber: `urn:uuid:${(0, uuid_1.v4)()}`,
            version: 1,
            metadata: {
                timestamp: finalResults.start,
                tools: [
                    {
                        vendor: 'RetireJS',
                        name: 'retire.js',
                        version: retire.version,
                    },
                ],
            },
            components: components,
        }, undefined, 2));
        writer.close(callback);
    };
}
exports.default = {
    configure: configureCycloneDXJSONLogger,
};
